Fixed some Secure Transport issues on OS X

swisspol opened the pull request:
(see updated commit message)

swisspol commented:
As noted in #3206 current libgit2 unit tests don't appear to test the Secure Transport code path. However the functional tests I have on top of libgit2 do test it during HTTPS cloning for instance.

carlosmn commented on src/stransport_stream.c:
```diff
@@ -114,14 +114,20 @@ static OSStatus write_cb(SSLConnectionRef conn, const void *data, size_t *len)
+	} while (left);
 
-	*len = ret;
+	*len = requested;
```
This change should not be necessary. The streams return when there is an error or when everything is written.

swisspol replied:
Good point, I just copy-pasted the implementation of `read_cb()` where it turns out it's not needed either.

carlosmn commented on src/stransport_stream.c:
```diff
@@ -150,9 +156,9 @@ static OSStatus read_cb(SSLConnectionRef conn, void *data, size_t *len)
+		if (ret <= 0) { /* There was an error or nothing could be read because stream is closed */
 			*len = 0;
-			return -1;
+			return errSSLClosedAbort;
```
If we read 0, it should mean that the other side closed. Why is it an abort?

swisspol replied:
Originally, I modified this function to behave like "read up to *len bytes from the connection" i.e. get rid of the `do ... while`. It worked the same way than the write function: "write up to *len bytes to the connection". In both cases if fewer bytes were handled, it would still return `noErr`.

But my functional tests failed when cloning a repo over HTTPS. So I figured, even though the docs do not mention that, maybe the behavior should be: "read or write this *exact* number of bytes to the connection or fail entirely if it can't". That's what you had done initially (why BTW?). I restored the `do ... while` in the read function and added the same behavior to the write function.

If you assume that's the desired behavior, and considering 0 can only be returned if fewer than `*len` were read, then the function should return an error. I picked `errSSLClosedAbort` for both read and write functions as it seemed the most sensitive among the possible SSL error codes.

swisspol replied:
BTW if you look at the use of this very API by Apple at (http://www.opensource.apple.com/source/Security/Security-55471/libsecurity_ssl/regressions/ssl-45-tls12.c), you will notice `SocketRead()` is returning an error when reading 0 bytes. And both `SocketRead()~ and `SocketWrite()` use loops to read / write exactly the passed-in number of bytes.

swisspol replied:
The above Apple link is actually from an older Security project version and is not around in the version shipping on 10.10. However, there's another use of the API at http://www.opensource.apple.com/source/Security/Security-55179.13/libsecurity_ssl/sslViewer/ioSock.c which has socket wrapper implementations of `SocketRead()` and `SocketWrite()` to be used with SSLSetIOFuncs().

swisspol replied:
Finally you can see how VLC does it here: https://github.com/videolan/vlc/blob/8a7b1417015346444b6d0a329a95a1dada535429/modules/misc/securetransport.c.

It does return `errSSLClosedGraceful` when reading 0, so a different behavior, and I would assume their code has been tested and deployed extensively.

This whole thing is pretty messy...

swisspol replied:
No matter what, all these uses of the API do loop for the write callback, so that's one thing to fix in libgit2.

swisspol commented:
*Moving line notes to actual discussion so that they don't disappear when updating the diff*

Originally, I modified this function to behave like "read up to *len bytes from the connection" i.e. get rid of the `do ... while`. It worked the same way than the write function: "write up to *len bytes to the connection". In both cases if fewer bytes were handled, it would still return `noErr`.

But my functional tests failed when cloning a repo over HTTPS. So I figured, even though the docs do not mention that, maybe the behavior should be: "read or write this *exact* number of bytes to the connection or fail entirely if it can't". That's what you had done initially (why BTW?). I restored the `do ... while` in the read function and added the same behavior to the write function.

If you assume that's the desired behavior, and considering 0 can only be returned if fewer than `*len` were read, then the function should return an error. I picked `errSSLClosedAbort` for both read and write functions as it seemed the most sensitive among the possible SSL error codes.

BTW if you look at the use of this very API by Apple at http://www.opensource.apple.com/source/Security/Security-55471/libsecurity_ssl/regressions/ssl-45-tls12.c, you will notice `SocketRead()` is returning an error when reading 0 bytes. And both `SocketRead()` and `SocketWrite()` use loops to read / write exactly the passed-in number of bytes.

The above Apple link is actually from an older Security project version and is not around in the version shipping on 10.10. However, there's another use of the API at http://www.opensource.apple.com/source/Security/Security-55179.13/libsecurity_ssl/sslViewer/ioSock.c which has socket wrapper implementations of `SocketRead()` and `SocketWrite()` to be used with SSLSetIOFuncs().

Finally you can see how VLC does it here: https://github.com/videolan/vlc/blob/8a7b1417015346444b6d0a329a95a1dada535429/modules/misc/securetransport.c.

It does return `errSSLClosedGraceful` when reading 0, so a different behavior, and I would assume their code has been tested and deployed extensively.

This whole thing is pretty messy...

PS: No matter what, all these uses of the API do loop for the write callback, so that's one thing to fix in libgit2.

swisspol commented:
OK, I've looked at a bunch of implementations of these callbacks (see comments above) and settled for this new version. See commit message for details.

carlosmn commented on src/stransport_stream.c:
```diff
@@ -110,20 +110,38 @@ int stransport_certificate(git_cert **out, git_stream *stream)
 
-	*len = ret;
+		off += ret;
+	} while (off < *len);
```
Again, this is already the behaviour we expect from our streams, so there is absolutely no need to loop over it. When `git_stream_write()` returns, it will have written all of the data it was given, or it will return an error.

carlosmn commented on src/stransport_stream.c:
```diff
@@ -110,20 +110,38 @@ int stransport_certificate(git_cert **out, git_stream *stream)
+	do {
+		ret = git_stream_write(io, data + off, *len - off, 0);
+		if (ret < 0) {
+			error = -36; /* "ioErr" from MacErrors.h */
```
Why use the magic number instead of the name?

swisspol replied:
That's what opensource.apple.com and VLC do. Importing `MacErrors.h` would require `#include <CoreServices/CoreServices.h>` which includes a ton of headers e.g. CarbonCore. Up to you.

swisspol replied:
Actually, here's the main reason: `MacErrors.h` is simply not available on iOS.

carlosmn commented on src/stransport_stream.c:
```diff
@@ -110,20 +110,38 @@ int stransport_certificate(git_cert **out, git_stream *stream)
+/*
+ * Contrary to typical network IO callbacks, Secure Transport write callback is
+ * expected to write *all* passed data, not just as much as it can, and any
+ * other case would be considered a failure.
```
This is not surprising in libgit2. Every write method will write everything or return a failure. It is surprising on the read side, but it's business as usual when writing.

swisspol commented:
All right, hopefully this is good now.
